File: cpp/src/protocol/opentrack_packet.cpp
#include "cameraunlock/protocol/opentrack_packet.h"
#include <cmath>
#include <cstring>

namespace cameraunlock {

bool OpenTrackPacket::TryParse(const void* data, size_t length, TrackingPose& pose) {
    if (data == nullptr || length < kMinPacketSize) {
        return false;
    }

    const auto* bytes = static_cast<const uint8_t*>(data);

    double yaw, pitch, roll;
    std::memcpy(&yaw, bytes + kYawOffset, sizeof(double));
    std::memcpy(&pitch, bytes + kPitchOffset, sizeof(double));
    std::memcpy(&roll, bytes + kRollOffset, sizeof(double));

    // Validate values are not NaN or Infinity
    if (std::isnan(yaw) || std::isinf(yaw) ||
        std::isnan(pitch) || std::isinf(pitch) ||
        std::isnan(roll) || std::isinf(roll)) {
        return false;
    }

    pose = TrackingPose(static_cast<float>(yaw), static_cast<float>(pitch), static_cast<float>(roll));
    return true;
}

bool OpenTrackPacket::TryParsePosition(const void* data, size_t length, PositionData& position) {
    if (data == nullptr || length < kMinPacketSize) {
        return false;
    }

    const auto* bytes = static_cast<const uint8_t*>(data);

    double px, py, pz;
    std::memcpy(&px, bytes + kPosXOffset, sizeof(double));
    std::memcpy(&py, bytes + kPosYOffset, sizeof(double));
    std::memcpy(&pz, bytes + kPosZOffset, sizeof(double));

    if (std::isnan(px) || std::isinf(px) ||
        std::isnan(py) || std::isinf(py) ||
        std::isnan(pz) || std::isinf(pz)) {
        return false;
    }

    // OpenTrack position is in centimeters, convert to meters
    position = PositionData(
        static_cast<float>(px * 0.01),
        static_cast<float>(py * 0.01),
        static_cast<float>(pz * 0.01)
    );
    return true;
}

bool OpenTrackPacket::TryParseAll(const void* data, size_t length, TrackingPose& pose, PositionData& position) {
    if (data == nullptr || length < kMinPacketSize) {
        return false;
    }

    const auto* bytes = static_cast<const uint8_t*>(data);

    double px, py, pz, yaw, pitch, roll;
    std::memcpy(&px, bytes + kPosXOffset, sizeof(double));
    std::memcpy(&py, bytes + kPosYOffset, sizeof(double));
    std::memcpy(&pz, bytes + kPosZOffset, sizeof(double));
    std::memcpy(&yaw, bytes + kYawOffset, sizeof(double));
    std::memcpy(&pitch, bytes + kPitchOffset, sizeof(double));
    std::memcpy(&roll, bytes + kRollOffset, sizeof(double));

    if (std::isnan(px) || std::isinf(px) ||
        std::isnan(py) || std::isinf(py) ||
        std::isnan(pz) || std::isinf(pz) ||
        std::isnan(yaw) || std::isinf(yaw) ||
        std::isnan(pitch) || std::isinf(pitch) ||
        std::isnan(roll) || std::isinf(roll)) {
        return false;
    }

    pose = TrackingPose(static_cast<float>(yaw), static_cast<float>(pitch), static_cast<float>(roll));
    position = PositionData(
        static_cast<float>(px * 0.01),
        static_cast<float>(py * 0.01),
        static_cast<float>(pz * 0.01)
    );
    return true;
}

}  // namespace cameraunlock

File: csharp/src/CameraUnlock.Core/Protocol/OpenTrackPacket.cs
using System;
using System.Runtime.InteropServices;
using CameraUnlock.Core.Data;

namespace CameraUnlock.Core.Protocol
{
    /// <summary>
    /// OpenTrack UDP packet parsing.
    /// OpenTrack sends 48 bytes of position/rotation data (6 doubles).
    /// Some versions also append 8 bytes for frame number.
    /// </summary>
    public static class OpenTrackPacket
    {
        /// <summary>Minimum packet size (6 doubles = 48 bytes).</summary>
        public const int MinPacketSize = 48;

        /// <summary>Byte offset of yaw in the packet.</summary>
        public const int YawOffset = 24;

        /// <summary>Byte offset of pitch in the packet.</summary>
        public const int PitchOffset = 32;

        /// <summary>Byte offset of roll in the packet.</summary>
        public const int RollOffset = 40;

        /// <summary>
        /// Attempts to parse an OpenTrack packet.
        /// </summary>
        /// <param name="data">Raw packet data.</param>
        /// <param name="pose">Parsed tracking pose if successful.</param>
        /// <returns>True if parsing succeeded.</returns>
        public static bool TryParse(byte[] data, out TrackingPose pose)
        {
            pose = default;

            if (data == null || data.Length < MinPacketSize)
            {
                return false;
            }

            double yaw = BitConverter.ToDouble(data, YawOffset);
            double pitch = BitConverter.ToDouble(data, PitchOffset);
            double roll = BitConverter.ToDouble(data, RollOffset);

            // Validate values are not NaN or Infinity
            if (double.IsNaN(yaw) || double.IsInfinity(yaw) ||
                double.IsNaN(pitch) || double.IsInfinity(pitch) ||
                double.IsNaN(roll) || double.IsInfinity(roll))
            {
                return false;
            }

            pose = new TrackingPose((float)yaw, (float)pitch, (float)roll);
            return true;
        }

#if NETSTANDARD2_1_OR_GREATER || NET5_0_OR_GREATER
        /// <summary>
        /// Attempts to parse an OpenTrack packet from a span.
        /// </summary>
        public static bool TryParse(ReadOnlySpan<byte> data, out TrackingPose pose)
        {
            pose = default;

            if (data.Length < MinPacketSize)
            {
                return false;
            }

            double yaw = BitConverter.ToDouble(data.Slice(YawOffset, 8));
            double pitch = BitConverter.ToDouble(data.Slice(PitchOffset, 8));
            double roll = BitConverter.ToDouble(data.Slice(RollOffset, 8));

            if (double.IsNaN(yaw) || double.IsInfinity(yaw) ||
                double.IsNaN(pitch) || double.IsInfinity(pitch) ||
                double.IsNaN(roll) || double.IsInfinity(roll))
            {
                return false;
            }

            pose = new TrackingPose((float)yaw, (float)pitch, (float)roll);
            return true;
        }
#endif
    }
}
